incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -741,30 +991,132 @@ private void onKafkaData(
+                    willFetchers.put(clientId, willFetcher);
+                }
             }
         }
 
-        protected abstract void handleKafkaData(DataFW data);
+        private void onKafkaFlush(
+            FlushFW flush)
+        {
+            final long sequence = flush.sequence();
+            final long acknowledge = flush.acknowledge();
 
-        protected abstract void onKafkaWindow(WindowFW window);
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
 
-        protected void onKafkaEnd(
-            EndFW end)
-        {
+            replySeq = sequence;
+
+            assert replyAck <= replySeq;
+
+            final OctetsFW extension = flush.extension();
+            final ExtensionFW flushEx = extension.get(extensionRO::tryWrap);
+            final KafkaFlushExFW kafkaFlushEx =
+                flushEx != null && flushEx.typeId() == kafkaTypeId ? extension.get(kafkaFlushExRO::tryWrap) : null;
+            final KafkaMergedFlushExFW kafkaMergedFlushEx =
+                kafkaFlushEx != null && kafkaFlushEx.kind() == KafkaDataExFW.KIND_MERGED ? kafkaFlushEx.merged() : null;
+            final Array32FW<KafkaOffsetFW> progress = kafkaMergedFlushEx != null ? kafkaMergedFlushEx.progress() : null;
+
+            if (progress != null)
+            {
+                final IntHashSet newPartitions = new IntHashSet();
+                progress.forEach(p -> newPartitions.add(p.partitionId()));
+                if (!newPartitions.equals(partitions))
+                {
+                    instanceId.regenerate();
+                    partitions = newPartitions;
+                }
+            }
+            else
+            {
+                //TODO: this means we lost all partitions during the assignment, so we need to regenerate, right?
+                instanceId.regenerate();
+            }
         }
 
-        protected void onKafkaFlush(
-            FlushFW flush)
+        private void onKafkaEnd(
+            EndFW end)
         {
+            final long sequence = end.sequence();
+            final long acknowledge = end.acknowledge();
+            final long traceId = end.traceId();
+            final long authorization = end.authorization();
+
+            assert acknowledge <= sequence;
+            assert sequence >= replySeq;
+
+            replySeq = sequence;
+            state = MqttKafkaState.closeReply(state);
+
+            assert replyAck <= replySeq;
+
+            doKafkaEnd(traceId, authorization);
+
+            if (reconnectDelay != 0)
+            {
+                if (reconnectAt != NO_CANCEL_ID)
+                {
+                    signaler.cancel(reconnectAt);
+                }
+
+                reconnectAt = signaler.signalAt(
+                    currentTimeMillis() + SECONDS.toMillis(reconnectDelay),
+                    SIGNAL_INITIATE_WILL_KAFKA_STREAM,
+                    MqttKafkaSessionFactory.this::onWillKafkaStreamInitializationSignal);
```

Comment:
Suggest hoisting this to a member field on the session factory, such as `handleReconnectWillStream`.